### ISSUE TITLE
regression in 2.2: crashes when working with certificates with empty issuerAltName

### ISSUE THREAD

**Issue by njsmith** [pyca/cryptography#4154]
Sample code:

```python
from cryptography import x509
from cryptography.hazmat.backends import default_backend

cert_pem = b"""
-----BEGIN CERTIFICATE-----
MIIDczCCAtygAwIBAgIBATANBgkqhkiG9w0BAQUFADCBgTELMAkGA1UEBhMCRkkx
DjAMBgNVBAgMBWR1bW15MQ4wDAYDVQQHDAVkdW1teTEOMAwGA1UECgwFZHVtbXkx
DjAMBgNVBAsMBWR1bW15MREwDwYDVQQDDAhTbmFrZU9pbDEfMB0GCSqGSIb3DQEJ
ARYQZHVtbXlAdGVzdC5sb2NhbDAeFw0xMTEyMjIwNzU4NDBaFw0yMTEyMTgwNzU4
NDBaMGExCzAJBgNVBAYTAkZJMQ4wDAYDVQQIDAVkdW1teTEOMAwGA1UEBwwFZHVt
bXkxDjAMBgNVBAoMBWR1bW15MQ4wDAYDVQQLDAVkdW1teTESMBAGA1UEAwwJbG9j
YWxob3N0MIGfMA0GCSqGSIb3DQEBAQUAA4GNADCBiQKBgQDXe3FqmCWvP8XPxqtT
+0bfL1Tvzvebi46k0WIcUV8bP3vyYiSRXG9ALmyzZH4GHY9UVs4OEDkCMDOBSezB
0y9ai/9doTNcaictdEBu8nfdXKoTtzrn+VX4UPrkH5hm7NQ1fTQuj1MR7yBCmYqN
3Q2Q+Efuujyx0FwBzAuy1aKYuwIDAQABo4IBGDCCARQwCQYDVR0TBAIwADAdBgNV
HQ4EFgQUG+dK5Uos08QUwAWofDb3a8YcYlIwgbYGA1UdIwSBrjCBq4AUGXd/I2Ji
QllF+3Wdx3NyBLszCi2hgYekgYQwgYExCzAJBgNVBAYTAkZJMQ4wDAYDVQQIDAVk
dW1teTEOMAwGA1UEBwwFZHVtbXkxDjAMBgNVBAoMBWR1bW15MQ4wDAYDVQQLDAVk
dW1teTERMA8GA1UEAwwIU25ha2VPaWwxHzAdBgkqhkiG9w0BCQEWEGR1bW15QHRl
c3QubG9jYWyCCQCz67HKL+G/4zAJBgNVHRIEAjAAMCQGA1UdEQQdMBuBDnJvb3RA
bG9jYWxob3N0gglsb2NhbGhvc3QwDQYJKoZIhvcNAQEFBQADgYEAgcW6X1ZUyufm
TFEqEAdpKXdL0rxDwcsM/qqqsXbkz17otH6ujPhBEagzdKtgeNKfy0aXz6rWZugk
lF0IqyC4mcI+vvfgGR5Iy4KdXMrIX98MbrvGJBfbdKhGW2b84wDV42DIDiD2ZGGe
6YZQQIo9LxjuOTf9jsvf+PIkbI4H0To=
-----END CERTIFICATE-----
"""

cert = x509.load_pem_x509_certificate(cert_pem, default_backend())

# Accessing this attribute crashes
cert.extensions
```

Running this with cryptography 2.2 (on Linux, from wheel), I get:

```
ValueError: Must supply at least one general name
```

With cryptography 2.1.4, it doesn't crash.

This certificate comes from [the urllib3 test suite](https://github.com/shazow/urllib3/blob/master/dummyserver/certs/server.crt), and apparently has an empty issuerAltName:

```
~/src/urllib3/dummyserver/certs$ openssl x509 -in server.crt -text
[...]
            X509v3 Issuer Alternative Name: 
                <EMPTY>
[...]
```

This is causing the urllib3 test suite to fail since 2.2 was released. Not sure if it should be considered an invalid cert that needs to be fixed, or a bug in cryptography.

**Comment by reaperhulk** [comment]
This is definitely an invalid cert by the RFC (which requires `   GeneralNames ::= SEQUENCE SIZE (1..MAX) OF GeneralName`). The PR that enforced this is  https://github.com/pyca/cryptography/pull/4128.

That said, failing on parsing junk certs has caused us quite a few headaches (#3856 and #3857 and now this). We've talked about a hazmat layer for x509, but I think the first step is to stop trying to enforce RFC compliance on parsing and instead enforce on generation.

The beginnings of that idea are implemented in https://github.com/pyca/cryptography/pull/3862 but I haven't revisited that in awhile. I guess that needs to be the next thing I look at to see if it's still an approach I like.

**Comment by citrus-it** [comment]
This has hit me too. Like the previous poster it was a bad certificate in our test suite that set `issuerAltName = issuer:copy` where the issuer did not have any `altSubjectNames`. In that case, OpenSSL generates an empty issuerAltName. We're fixing our test suite but it seems that this might be a fairly common flavour of non-RFC compliant certificate.

**Comment by njsmith** [comment]
This was fixed by #4161 and #4163, and the fix is released in v2.2.1, so closing.
